py4j-python/src/py4j/java_gateway.py

```diff
@@ -678,6 +760,39 @@ def __getattr__(self, name):
         # The name is a method
         return self._methods[name]
 
+    def __dir__(self):
+        self._populate_fields()
+        return list(set(self._methods.keys()) | self._field_names)
+
+    def _populate_fields(self):
+        if not self._fully_populated:
```

Comment:
I am going to change the code to be a single tuple holding both the sequence and cache so that I don't have to worry about them getting out of step.